### PR TITLE
Allow AnsibleVaultEncryptedUnicode to pass through exec_jsonrpc

### LINKED ISSUE #47200: ansible-connection complaining about vaulted ansible_ssh_pass not being JSON-serializable
##### SUMMARY
After upgrading to ansible 2.7.0, `ansible-connection` is complaining about my vaulted `ansible_ssh_pass` not being JSON-serializable. Converting the password to plaintext in my `group_vars` file allows the playbook to run.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

slxos_facts

##### ANSIBLE VERSION

```
2.7.0
```

##### CONFIGURATION

```
ANSIBLE_HOST_KEY_CHECKING=False
#ANSIBLE_DEBUG=True
#ANSIBLE_LOG_PATH=/tmp/playbook/log/ansible.log
ANSIBLE_PERSISTENT_CONNECT_TIMEOUT=60
ANSIBLE_PERSISTENT_COMMAND_TIMEOUT=30
#ANSIBLE_STDOUT_CALLBACK=json
```

##### OS / ENVIRONMENT
using ansible control in a Docker container on Debian 9

##### STEPS TO REPRODUCE
Run `slxos_facts` against any SLX device

```
---
- hosts: slx
  gather_facts: yes

  tasks:

  - name: get interface info
    slxos_facts:
      gather_subset: interfaces
    register: result

  - debug: var=result
```

##### EXPECTED RESULTS

Interface details would be gathered from my list of SLX devices.

##### ACTUAL RESULTS

```
fatal: [redacted]: FAILED! => {"msg": "Traceback (most recent call last):
  File \"/usr/bin/ansible-connection\", line 321, in <module>
    main()
  File \"/usr/bin/ansible-connection\", line 287, in main
    conn.set_options(var_options=variables)
  File \"/usr/lib/python2.7/site-packages/ansible/module_utils/connection.py\", line 167, in __rpc__
    response = self._exec_jsonrpc(name, *args, **kwargs)
  File \"/usr/lib/python2.7/site-packages/ansible/module_utils/connection.py\", line 144, in _exec_jsonrpc
    data = json.dumps(req)
  File \"/usr/lib/python2.7/json/__init__.py\", line 244, in dumps
    return _default_encoder.encode(obj)
  File \"/usr/lib/python2.7/json/encoder.py\", line 207, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File \"/usr/lib/python2.7/json/encoder.py\", line 270, in iterencode
    return _iterencode(o, 0)
  File \"/usr/lib/python2.7/json/encoder.py\", line 184, in default
    raise TypeError(repr(o) + \" is not JSON serializable\")
TypeError: u'***redacted***' is not JSON serializable
"}
```

```
docker run -it --rm -v $PWD:/tmp/playbook:Z --env-file ansible.env ansiblecm brcd_sflow.yml --vault-password-file vault/vault_pass -i $BRCD_ANS
```

### PR DESCRIPTION
**Pull request by Qalthos**
##### SUMMARY

Fixes #47200 

~~Before I start copying over the rest of AnsibleJSONEncoder, does anyone have a better idea of how to get that functionality into module_utils?~~

I just duplicated AnsibleJSONEncoder into connection. Could probably move it somewhere else.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

module_utils/connection.py

##### ANSIBLE VERSION

```paste below
2.8
```

### CODE REVIEW

File: lib/ansible/module_utils/connection.py [hunk 1]
```diff
@@ -146,21 +146,26 @@ def _exec_jsonrpc(self, name, *args, **kwargs):
...
-                            "Please open an issue and mention that the culprit is most likely '%s'" % key
-                        )
-
-            raise ConnectionError(
-                "Failed to encode some variables as JSON for communication with ansible-connection. "
-                "The original exception was: %s" % to_text(exc)
-            )
+                options = data.get('var_options', {})
+
+                for key, value in iteritems(options):
```
Comment (bcoca) [inline lib/ansible/module_utils/connection.py]: This works when he vault is on the top level, but vaults can appear at any point int the data structure:
```yaml
nonvault:
   key1: stillnovaul
   key2:
       otherdict: otherval
       thisisvault: !vault ...
```

**Comment by joefodor** [comment]
I am having the same issue on 2.7.5

**Comment by Im0** [comment]
@joefodor looks like this was fixed for 2.8 only so far.  It's not made it into 2.7.7 as I'm still seeing the serializable issue when trying to decrypt. :(  Really hoping that this can be backported to 2.7.x.  What do you think @Qalthos ?

**Comment by epicanthal** [comment]
What is the workaround for 2.7.x?  After installing 2.7.x, copy connection.py from devel branch?

**Comment by Im0** [comment]
Hi @epicanthal - I've not tested this as I'd prefer to see it backported, however, it appears that @ericdost had success with what you suggest in:  https://github.com/ansible/ansible/issues/47200

**Pull request merged**